service/matching/matchingEngine.go

```diff
@@ -153,20 +153,23 @@ func (e *matchingEngineImpl) String() string {
 	return r
 }
 
-// Returns taskListManager for a task list. If not already cached gets new range from DB and if successful creates one.
 func (e *matchingEngineImpl) getTaskListManager(taskList *taskListID) (taskListManager, error) {
-	e.taskListsLock.RLock()
-	if result, ok := e.taskLists[*taskList]; ok {
-		e.taskListsLock.RUnlock()
-		return result, nil
-	}
-	e.taskListsLock.RUnlock()
-	mgr := newTaskListManager(e, taskList, e.config)
+	return e.getTaskListManagerWithRPS(taskList, nil)
+}
+
+// Returns taskListManager for a task list. If not already cached gets new range from DB and
+// if successful creates one. The passed in throttling limit determines how many tasks are
+// dispatcher per second.
+func (e *matchingEngineImpl) getTaskListManagerWithRPS(
+	taskList *taskListID, maxDispatchPerSecond *float64,
+) (taskListManager, error) {
 	e.taskListsLock.Lock()
 	if result, ok := e.taskLists[*taskList]; ok {
+		result.UpdateMaxDispatch(maxDispatchPerSecond)
```

Comment:
There is one way with
RLock
Check map, 
if ok, 
  runlock, update max dispatch
if not ok,
  RUnlock
  WLock
  Again check map
  if still not ok, then lock and write

So we will have some duplicate code, we will be checking if it is in the map twice. That way, if it is  in map originally, it will onyl write lock. Makes code uglier but avoids write lock in majority of the cases.

Reply:
Removed Update method but FYI most of this code is still needed for task list creation.